### PR TITLE
KAFKA-5282: Use a factory method to create producers/consumers and close them in tearDown

### PR DESCRIPTION
**Pull request by vahidhashemian**

### CODE REVIEW

**Comment by vahidhashemian** [comment]
@apurvam @ijuma For your review. Thanks.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala [hunk 1]
```diff
@@ -151,208 +165,182 @@ class TransactionsTest extends KafkaServerTestHarness {
...
   }
 
   @Test
   def testFencingOnCommit() = {
-    val transactionalId = "my-t.id"
-    val producer1 = TestUtils.createTransactionalProducer(transactionalId, servers)
-    val producer2 = TestUtils.createTransactionalProducer(transactionalId, servers)
-    val consumer = transactionalConsumer()
+    val producer1 = transactionalProducers.head
+    val producer2 = transactionalProducers.tail.head
```
Comment (apurvam) [inline core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: Can't we retrieve by the index here?

**Review by apurvam — COMMENTED** [review]
Thanks for the patch. It looks good to me overall, apart from one minor comment.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala (hunk 1 above)
Comment (vahidhashemian) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: Thanks for reviewing. We could retrieve by index too. I'll change `.tail.head` to `.get(1)`. I hope that works.

**Review by apurvam — APPROVED** [review]
Thanks for the patch, LGTM. 

I would prefer `get(0)`, and `get(1)`  --rather than `head` and `get(1)`-- for the sake of consistency. But it's up to you whether you want to make that change. 

(Note that a committer will also need to approve and merge this).

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala [hunk 2]
```diff
@@ -151,208 +165,182 @@ class TransactionsTest extends KafkaServerTestHarness {
...
-      val records = pollUntilExactlyNumRecords(consumer, 2)
-      records.zipWithIndex.foreach { case (record, i) =>
-        TestUtils.assertCommittedAndGetValue(record)
-      }
-    } finally {
-      consumer.close()
-      producer1.close()
-      producer2.close()
+    val records = pollUntilExactlyNumRecords(consumer, 2)
+    records.zipWithIndex.foreach { case (record, i) =>
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: Why is there a `zipWithIndex` here? We're not using `i`.

Comment (apurvam) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: That's actually from the original code. I copied the pattern from another test (can't remember which). Incidentally, I could not get the `zip` to work at the time, though I think I would if I tried now. (It's amazing how a couple of weeks of intensive scala stimulates the imagination).

Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: I think `foreach` is enough here. I am guessing in the other case, the index was being used to retrieve an item from another collection and `zip` would indeed make sense then.

Comment (vahidhashemian) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: I removed this one and other ones like this.

**Comment by hachikuji** [comment]
@vahidhashemian Thanks for the patch. Please rebase when you have a chance.

**Comment by vahidhashemian** [comment]
@hachikuji I rebased the patch. Meant to do the rebase in a new commit, but missed a step. I hope it's not too inconvenient. Thanks.

**Review by ijuma — APPROVED** [review]
Thanks, LGTM. I fixed a few things before merging to trunk and 0.11.0:

1. Fixed merge conflicts
2. Removed a number of redundant `case`
3. Removed a number of redundant `()`
4. Replaced unused variables with `_`
5. Fixed usage of `:` so that there is no space before the colon

**Pull request closed without merging**